serverAuth caches first inferred baseURL across hosts

## Summary
`serverAuth()` caches the first inferred `baseURL` globally, so subsequent requests on different hosts keep using the first host.

## Why this matters
When `runtimeConfig.public.siteUrl` is not set (host inferred from request headers), multi-host deployments can generate incorrect callback URLs and auth links after the first request initializes the singleton.

## Evidence
- Singleton cache is initialized once and reused:
  - `/Users/maxi/nuxt/better-auth/src/runtime/server/utils/auth.ts:13`
  - `/Users/maxi/nuxt/better-auth/src/runtime/server/utils/auth.ts:148`
- `baseURL` is inferred from request origin before creating the singleton:
  - `/Users/maxi/nuxt/better-auth/src/runtime/server/utils/auth.ts:153`

## Reproduction
Added a focused regression test:
- `/Users/maxi/nuxt/better-auth/test/server-auth-base-url-cache.test.ts`
- Fixture route: `/Users/maxi/nuxt/better-auth/test/cases/base-url-inference/server/api/test/base-url.get.ts`

Run:
```bash
pnpm test test/server-auth-base-url-cache.test.ts
```

<details>
<summary>Explicit failing test output</summary>

```txt
FAIL  test/server-auth-base-url-cache.test.ts > serverAuth baseURL inference > derives baseURL from each request host when siteUrl is unset
AssertionError: expected 'https://first.example.com' to be 'https://second.example.com'

Expected: "https://second.example.com"
Received: "https://first.example.com"

❯ test/server-auth-base-url-cache.test.ts:30:32
    29| expect(firstBody.baseURL).toBe('https://first.example.com')
    30| expect(secondBody.baseURL).toBe('https://second.example.com')
```

</details>

## Expected behavior
When `siteUrl` is unset and request-origin inference is used, base URL should reflect the current request host (or there should be a deterministic per-request strategy that avoids stale host reuse).

## Acceptance criteria
- The regression test passes.
- Host-derived base URL does not persist incorrectly across different requests.
- Behavior is documented for multi-host deployments when `siteUrl` is unset.

--- a/src/runtime/server/utils/auth.ts
+++ b/src/runtime/server/utils/auth.ts
@@ -10,7 +10,7 @@ import { withoutProtocol } from 'ufo'
 
 type AuthInstance = Auth<ReturnType<typeof createServerAuth>>
 
-let _auth: AuthInstance | null = null
+const _authCache = new Map<string, AuthInstance>()
 let _baseURLInferenceLogged = false
 
 function normalizeLoopbackOrigin(origin: string): string {
@@ -144,18 +144,21 @@ function getBaseURL(event?: H3Event): string {
   throw new Error('siteUrl required. Set NUXT_PUBLIC_SITE_URL.')
 }
 
-/** Returns Better Auth instance. Pass event for accurate URL detection on first call. */
+/** Returns Better Auth instance. Caches per resolved host (or single instance when siteUrl is explicit). */
 export function serverAuth(event?: H3Event): AuthInstance {
-  if (_auth)
-    return _auth
-
   const runtimeConfig = useRuntimeConfig()
   const siteUrl = getBaseURL(event)
+  const hasExplicitSiteUrl = runtimeConfig.public.siteUrl && typeof runtimeConfig.public.siteUrl === 'string'
+  const cacheKey = hasExplicitSiteUrl ? '__explicit__' : siteUrl
+
+  const cached = _authCache.get(cacheKey)
+  if (cached)
+    return cached
 
   const database = createDatabase()
   const userConfig = createServerAuth({ runtimeConfig, db })
 
-  _auth = betterAuth({
+  const auth = betterAuth({
     ...userConfig,
     ...(database && { database }),
     secondaryStorage: createSecondaryStorage(),
@@ -163,5 +166,6 @@ export function serverAuth(event?: H3Event): AuthInstance {
     baseURL: siteUrl,
   })
 
-  return _auth
+  _authCache.set(cacheKey, auth)
+  return auth
 }

--- a/test/cases/base-url-inference/.nuxtrc
+++ b/test/cases/base-url-inference/.nuxtrc
@@ -1,0 +1,1 @@
+setups.@onmax/nuxt-better-auth="0.0.2-alpha.21"

--- a/test/cases/base-url-inference/app/app.vue
+++ b/test/cases/base-url-inference/app/app.vue
@@ -1,0 +1,3 @@
+<template>
+  <NuxtPage />
+</template>

--- a/test/cases/base-url-inference/app/auth.config.ts
+++ b/test/cases/base-url-inference/app/auth.config.ts
@@ -1,0 +1,3 @@
+import { defineClientAuth } from '../../../../src/runtime/config'
+
+export default defineClientAuth({})

--- a/test/cases/base-url-inference/nuxt.config.ts
+++ b/test/cases/base-url-inference/nuxt.config.ts
@@ -1,0 +1,6 @@
+export default defineNuxtConfig({
+  modules: ['../../../src/module'],
+  runtimeConfig: {
+    betterAuthSecret: 'test-secret-for-testing-only-32chars!',
+  },
+})

--- a/test/cases/base-url-inference/server/api/test/base-url.get.ts
+++ b/test/cases/base-url-inference/server/api/test/base-url.get.ts
@@ -1,0 +1,4 @@
+export default defineEventHandler((event) => {
+  const auth = serverAuth(event) as { options?: { baseURL?: string } }
+  return { baseURL: auth.options?.baseURL }
+})

--- a/test/cases/base-url-inference/server/auth.config.ts
+++ b/test/cases/base-url-inference/server/auth.config.ts
@@ -1,0 +1,11 @@
+import { defineServerAuth } from '../../../../src/runtime/config'
+
+export default defineServerAuth(({ runtimeConfig }) => ({
+  appName: 'Base URL inference test app',
+  socialProviders: {
+    github: {
+      clientId: runtimeConfig.github?.clientId || 'test',
+      clientSecret: runtimeConfig.github?.clientSecret || 'test',
+    },
+  },
+}))

--- a/test/server-auth-base-url-cache.test.ts
+++ b/test/server-auth-base-url-cache.test.ts
@@ -1,0 +1,32 @@
+import { fileURLToPath } from 'node:url'
+import { setup, url } from '@nuxt/test-utils/e2e'
+import { describe, expect, it } from 'vitest'
+
+describe('serverAuth baseURL inference', async () => {
+  await setup({
+    rootDir: fileURLToPath(new URL('./cases/base-url-inference', import.meta.url)),
+  })
+
+  it('derives baseURL from each request host when siteUrl is unset', async () => {
+    const firstResponse = await fetch(url('/api/test/base-url'), {
+      headers: {
+        'x-forwarded-host': 'first.example.com',
+        'x-forwarded-proto': 'https',
+      },
+    })
+    expect(firstResponse.status).toBe(200)
+    const firstBody = await firstResponse.json() as { baseURL: string | undefined }
+
+    const secondResponse = await fetch(url('/api/test/base-url'), {
+      headers: {
+        'x-forwarded-host': 'second.example.com',
+        'x-forwarded-proto': 'https',
+      },
+    })
+    expect(secondResponse.status).toBe(200)
+    const secondBody = await secondResponse.json() as { baseURL: string | undefined }
+
+    expect(firstBody.baseURL).toBe('https://first.example.com')
+    expect(secondBody.baseURL).toBe('https://second.example.com')
+  })
+})

